intermediate_source/parametrizations.py

```diff
@@ -0,0 +1,353 @@
+
+    def right_inverse(self, A):
+        if not self.is_skew(A):
+            raise ValueError(f"The provided matrix {A} is not skew-symmetric")
+        return A.triu(1)
+
+###############################################################################
+# We may now initialize a layer that is parametrized with ``Skew``
+layer = nn.Linear(3, 3)
+P.register_parametrization(layer, "weight", Skew())
+X = torch.rand(3, 3)
+X = X - X.T                             # X is now skew-symmetric
+layer.weight = X                        # Initialize layer.weight to be X
+assert torch.allclose(layer.weight, X)  # layer.weight == X
+
+###############################################################################
+# This ``right_inverse`` works as expected when we compose parametrizations. To see this, let's
+# upgrade the Cayley parametrization to also support being initialized
+class CayleyMap(nn.Module):
+    def __init__(self, n):
+        self.register_buffer("Id", torch.eye(n))
+
+    def forward(self, X):
+        # (I + X)(I - X)^{-1}
+        return torch.solve(self.Id + X, self.Id - X).solution
+
+    def right_inverse(self, A):
+        # See https://en.wikipedia.org/wiki/Cayley_transform#Matrix_map
+        # (X - I)(X + I)^{-1}
+        return torch.solve(X - self.Id, self.Id + X).solution
+
+layer_orthogonal = nn.Linear(3, 3)
+P.register_parametrization(layer_orthogonal, "weight", Skew())
+P.register_parametrization(layer_orthogonal, "weight", CayleyMap(3))
+# Sample an orthogonal matrix with positive determinant
+X = torch.empty(3, 3)
+nn.init.orthogonal_(X)
+if X.det() < 0.:
+    X[0].neg_()
+layer_orthogonal.weight = X
+assert torch.allclose(X, layer_orthogonal.weight)  # layer_orthogonal.weight == X
+
+###############################################################################
+# This initialization step can be written more succinctly as
+layer_orthogonal.weight = nn.init.orthogonal_(layer_orthogonal.weight)
+
+###############################################################################
+# The name of this method comes from the fact that we would often expect
+# that ``forward(right_inverse(X)) == X``. This is a direct way of rewriting that
+# the forward afer the initalization with value ``X`` should return the value ``X``.
+# This constraint is not enforced in the code. In fact, at times, it might be of
+# interest to relax this relation. For example, consider the following implementation
+# of a randomized pruning method:
+class PruningParametrization(nn.Module):
+    def __init__(self, X, p_drop=0.2):
+        # sample zeros with probability p_drop
+        mask = torch.full_like(X, 1.0 - p_drop)
+        self.mask = torch.bernoulli(mask)
+
+    def forward(self, X):
+        return X * self.mask
+
+    def right_inverse(self, A):
+        return A
+
+###############################################################################
+# In this case, it is not true that ``forward(right_inverse(X)) == X``. This is
```

Comment:
Would reparametrizations work for double-backward? A related question asked here: https://github.com/pytorch/pytorch/issues/55368

Reply:
That is a good point. At the moment that example but using
```python
import torch.nn.utils.parametrize as P
model = P.register_parametrization(torch.nn.Linear(2, 5), "weight", torch.nn.ReLU())
```
also breaks with the same error message. That being said, it smells like it should be a problem in the implementation of parametrizations. Perhaps @albanD can give a bit more insight on what's going on..

Reply:
There is no reason it wouldn't.
Your implementation of the reparametrization will need to be double differentiable though.

Reply:
I guess it's also worth an explicit discussion somewhere

Reply:
Not sure what you mean about the relu part though. This works fine on colab for me:
```python
# !pip uninstall --y torch
# !pip install --pre torch -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html

import torch
import torch.nn.utils.parametrize as P
model = P.register_parametrization(torch.nn.Linear(2, 5), "weight", torch.nn.ReLU())


inp = torch.rand(1, 2, requires_grad=True)

out = model(inp)

g = torch.autograd.grad(out.sum(), model.parameters(), create_graph=True)

print(g)

g[1].exp().sum().backward()

print(inp.grad)
```

Reply:
I meant that this example (which is a modification of that in that PR) breaks:
```python
import torch
import torch.nn.utils.parametrize as P
model = P.register_parametrization(torch.nn.Linear(2, 5), "weight", torch.nn.ReLU())
opt1 = torch.optim.SGD(model.parameters(), lr=1e-3)
opt2 = torch.optim.SGD(model.parameters(), lr=1e-3)

output = model(torch.randn(7, 2))
loss = output.abs().mean()

opt1.zero_grad(); loss.backward(retain_graph=True); opt1.step() # first propagation
opt2.zero_grad(); loss.backward(); opt2.step() # second
```
but it might be because `ReLU` is not twice differentiable?